beacon-chain/sync/data_columns.go

```diff
@@ -0,0 +1,444 @@
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/interfaces"
+	eth "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v5/runtime/logging"
+	"github.com/prysmaticlabs/prysm/v5/runtime/version"
+	"github.com/sirupsen/logrus"
+)
+
+// RequestDataColumnSidecars sends a data column sidecars by root request to one
+// or more peers that can provide the needed data columns.
+func RequestDataColumnSidecars(
+	ctx context.Context,
+	dataColumns map[uint64]bool,
+	block interfaces.ReadOnlySignedBeaconBlock,
+	blkRoot [32]byte,
+	peers []core.PeerID,
+	clock *startup.Clock,
+	p2p p2p.P2P,
+	ctxMap ContextByteVersions,
+	newColumnsVerifier verification.NewDataColumnsVerifier,
+) ([]blocks.RODataColumn, error) {
+	if len(dataColumns) == 0 {
+		return nil, nil
+	}
+
+	// Assemble the peers who can provide the needed data columns.
+	dataColumnsByAdmissiblePeer, _, _, err := AdmissiblePeersForDataColumns(peers, dataColumns, p2p)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get admissible peers for data columns")
+	}
+
+	sidecars := make([]blocks.RODataColumn, 0, len(dataColumns))
+	remainingColumns := make(map[uint64]bool, len(dataColumns))
+	for col := range dataColumns {
+		remainingColumns[col] = true
+	}
+
+	for len(dataColumnsByAdmissiblePeer) > 0 {
+		// Try to select peers excluding bad peers
+		peersToFetchFrom, err := SelectPeersToFetchDataColumnsFrom(remainingColumns, dataColumnsByAdmissiblePeer)
+		if err != nil {
+			// Return an error if some columns are unavailable from the filtered set
+			// of peers. Filtering out bad peers can make columns unavailable, and
+			// when that happens, the caller needs to know about it.
+			return nil, errors.Wrap(err, "couldn't select peers to fetch data columns from")
+		}
+
+		// Request the data columns from each peer
+		successfulColumns := make(map[uint64]bool)
+		for peer, dataColumns := range peersToFetchFrom {
+			request, err := RequestsForDataColumnsByRoot(blkRoot, dataColumns)
+			if err != nil {
+				log.WithError(err).Debug("Failed to build request for data columns")
+				continue
+			}
+
+			peerSidecars, err := SendDataColumnSidecarsByRootRequest(ctx, clock, p2p, peer, ctxMap, &request)
+			if err != nil {
+				// Remove this peer since it failed to respond correctly
+				delete(dataColumnsByAdmissiblePeer, peer)
+				log.WithFields(logrus.Fields{
+					"peer":      peer.String(),
+					"blockRoot": fmt.Sprintf("%#x", blkRoot),
+					"error":     err.Error(),
+				}).Debug("Failed to request data columns from peer")
+				continue
+			}
+
+			// Mark columns as successful
+			for _, sidecar := range peerSidecars {
+				colIndex := sidecar.ColumnIndex
+				successfulColumns[colIndex] = true
+			}
+
+			for _, colIndex := range dataColumns {
+				if !successfulColumns[colIndex] {
+					// Remove this peer if any requested column wasn't successful
```

Comment:
The case where a peer did not respond all its request column is not tested.
I think it's quite important to test it as well.

Reply:
[a93bdf7](https://github.com/prysmaticlabs/prysm/pull/14977/commits/a93bdf7ac2a69f8e2dadf8ca2e8f7c55ce501a48)